src/services/indexedCollection.ts

```diff
@@ -0,0 +1,99 @@
+import { createSequence } from "@/createSequence"
+
+type IndexedCollectionId = number
+type CollectionData<TData> = Map<IndexedCollectionId, TData>
+type CollectionIndex<TData, TKey extends keyof TData> = Map<TData[TKey], Set<IndexedCollectionId>>
+
+type IndexedCollection<TData, TKeys extends keyof TData> = {
+  addItem: (item: TData) => void
+  deleteItem: <TKey extends TKeys>(index: TKey, value: TData[TKey]) => void
+  findItem: <TKey extends TKeys>(index: TKey, value?: TData[TKey]) => TData[]
```

Comment:
nit: 
I'd name these `deleteItems` and `getItems` which I think more accurately describes what they do. `get` rather than `find` because its a direct access rather than a search/loop. `items` because it will delete or get multiple.